[Triton] [Inductor] Enable TMA store for TMA mm templates

njriasan opened the pull request:
Summary:
Adds support for TMA store in all TMA matmul templates (notably persistent_tma including addmm and scaled_mm). This works by requiring a template be registered with `tma_store=True` and when met constructs indices/range_trees to hook into the existing code base's TMA store support.

This also includes a couple notable changes:
- Adds support in the TMA template support for checking the output layout.
- Adds support for "hoisting" the tensor descriptor to the top of the kernel. This will currently only be used by template code right now, but in principle it can be generalized to other implementation.
- Supports considering multiple indices as the "contiguous" index. This is handled with support for transposing the input data when the alignment is no longer consistent. In general since the TMA support is derived from the index it doesn't seems reasonable that the 1D index math forces a certain alignment depending on index ordering so long as the layout matches.

Test Plan:
Tested with test_max_autotune.py unit tests.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

njriasan commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1018,26 +1093,74 @@ def store_output(
+
+                assert len(tma_info.block_sizes) == len(output_index), "Unexpected number of indices"
+                assert len(tma_info.shapes) == len(output_index), "Unexpected number of indices"
+                # TODO: Fix all of the naming logic.
```
I need to cleanup this code a lot, but the biggest component here is that I'm generating the remapping from TMA indices to regular pointer indices (happy to split this off into a helper). To do this I'm generating a lot of intermediate names and I'm not familiar with the ideal way of doing that in this code base (a lot of the helpers seems to have side effects).

I'll keep looking around/ask an llm, but any pointers are appreciates (cc: @eellison). This code is now fully working for fusions in addmm.

njriasan commented on torch/_inductor/utils.py:
```diff
@@ -1665,11 +1667,31 @@ def can_use_tma(*matrices: IRNode, add_guards: bool = False) -> bool:
+        strides = layout.stride
+        dtype = layout.dtype
+
+        # TODO: Check Base pointer 16-byte aligned for output?
```
Please let me know if you have thoughts on how I can check this. Also happy to do this as a followup.

eellison replied:
Check if `layout.offset` is statically divisible by 16

eellison reviewed (commented):
Could we add this logic to [triton codegen store](https://github.com/pytorch/pytorch/blob/179511694cbbc22be87e5fa4b6965e41129032d1/torch/_inductor/codegen/triton.py#L2609) instead ? 

Potentially with a flag on the kernel to prefer tma stores.

njriasan commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2304,31 +2355,44 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
+                            not tma_compatibility_checker.are_block_parameters_compatible(
+                                block_params
+                            )
+                            and copy_shape is not None
```
This may not be a sufficient check that we know val's shape.

NikhilAPatel reviewed (approved):
Just had one comment above about the decision to enable TMA store by default, but otherwise looks good to me! Very impressive work thanks for tackling this!

njriasan commented on torch/_inductor/utils.py:
```diff
@@ -1679,23 +1681,44 @@ def can_use_tma(*matrices: IRNode, add_guards: bool = False) -> bool:
-        # Base pointer 16-byte aligned
-        if x.get_name() in V.graph.unaligned_buffers:
+        if dtype not in (torch.float16, torch.bfloat16, torch.float8_e4m3fn) and (
+            not allow_float32 or dtype != torch.float32
```
allow_float32 shouldn't be necessary in general, but this PR has already gotten quite large, so I want to reduce the chance of a revert. This is necessary because some of the fp8 examples have float32 outputs. In general I see no reason float32 shouldn't be supported.

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2304,31 +2356,49 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
+                        block_params
                     ):
-                        return None
+                        # Some inputs may intend to store the values as (X, Y),
+                        # not (Y, X). As a result, try and tranpsose the blocks
```
Do you have any examples of this ? And when they would be tma compatible as Y, X but not X, Y ?

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2451,23 +2523,45 @@ def codegen_block_ptr(
-        ):
-            if V.graph.sizevars.statically_known_equals(dim, broadcast_dim):
-                indexing.broadcasting_dims[idx] = False
+        # TMA stores may require transposing the data to ensure we are contiguous along
+        # the final dimension. We do this by checking the shape information on value.
+        # It can either
+        #    1. Match the final shape. In this case no broadcast/reshape
```
Is this different from other triton stores ? In triton, you can store [X, Y] and it can be coalesced on either dimension and you do not need to manually transpose.

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1045,26 +1122,72 @@ def store_output(
-            output_index = self.rename_indexing(output_index)
-            if output_index == contiguous_index:
-                output_index = sympy.Symbol("xindex", integer=True)
+            if self.tma_store:
```
Couldnt this be a flag on the template?

eellison reviewed (commented):
Looks good, a few questions. Also cc @blaine-rister for block pointer review

njriasan commented:
@eellison I applied your feedback and @NikhilAPatel's feedback to disable the TMA_STORE by default. @NikhilAPatel now the same code works for TMA and no TMA, so you can now autotune over the TMA_STORE.

blaine-rister reviewed (commented):
Very cool and useful PR. The existing logic seems reasonable, but I left a few nits about opportunities to consolidate the code, or edge cases we might want to explore. Also, it might be possible to generalize this to ND without much added complexity. See the detailed comments above.

cc @kundaMwiza  who implemented automatic TMA codegen.

njriasan reviewed (commented):
@blaine-rister Thanks for the feedback! I think we should do the ND generalization in a followup PR, but overall I agree with and appreciate the feedback.

njriasan committed:
[Triton] [Inductor] Enable TMA store for TMA mm templates (#160480)

Summary:
Adds support for TMA store in all TMA matmul templates (notably persistent_tma including addmm and scaled_mm). This works by requiring a template be registered with `tma_store=True` and when met constructs indices/range_trees to hook into the existing code base's TMA store support.

This also includes a couple notable changes:
- Adds support in the TMA template support for checking the output layout.
- Adds support for "hoisting" the tensor descriptor to the top of the kernel. This will currently only be used by template code right now, but in principle it can be generalized to other implementation.
- Supports considering multiple indices as the "contiguous" index. This is handled with support for transposing the input data when the alignment is no longer consistent. In general since the TMA support is derived from the index it doesn't seems reasonable that the 1D index math forces a certain alignment depending on index ordering so long as the layout matches.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/160480

Test Plan:
Tested with test_max_autotune.py unit tests.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

Rollback Plan:

Differential Revision: D82313190

Pulled By: njriasan

njriasan committed:
Merge branch 'main' into export-D79378362

njriasan committed:
Fix broken test

njriasan committed:
Apply mixin feedback

njriasan committed:
Apply feedback on cleaning up tma check

njriasan committed:
fixed linting

njriasan committed:
Fully fixed linting for edited files

njriasan committed:
Restrict transpose for now

njriasan committed:
Fix CI

njriasan committed:
Removed restriction

njriasan committed:
Removed asserts until code is cleaned up

njriasan commented:
@pytorchbot merge
